src/appleseed/renderer/kernel/lighting/volumelightingintegrator.cpp

```diff
@@ -0,0 +1,601 @@
+                    extinction_coef,
+                    sampling_context,
+                    mis_heuristic,
+                    radiance,
+                    false);
+            }
+        }
+    }
+
+    // Add contributions from the light set.
+    if (m_light_sampler.has_lightset())
+    {
+        sampling_context.split_in_place(3, m_distance_sample_count);
+
+        for (size_t i = 0; i < m_distance_sample_count; ++i)
+        {
+            // Sample the light set.
+            sampling_context.split_in_place(3, 1);
+            LightSample light_sample;
+            m_light_sampler.sample_lightset(
+                m_time,
+                sampling_context.next2<Vector3f>(),
+                m_shading_point,
+                light_sample);
+
+            for (size_t j = 0; j < m_light_sample_count; ++j)
+                add_single_distance_sample_contribution(
+                    &light_sample,
+                    extinction_coef,
+                    sampling_context,
+                    mis_heuristic,
+                    radiance,
+                    false);
+
+            if (!ScatteringMode::has_volume(m_scattering_modes))
+                add_single_distance_sample_contribution(
+                    &light_sample,
+                    extinction_coef,
+                    sampling_context,
+                    mis_heuristic,
+                    radiance,
+                    true);
+        }
+    }
+}
+
+void VolumeLightingIntegrator::compute_radiance_exponential_sampling_only(
+    SamplingContext&            sampling_context,
+    const MISHeuristic          mis_heuristic,
+    ShadingComponents&          radiance) const
+{
+    radiance.set(0.0f);
+
+    // No light source in the scene.
+    if (!m_light_sampler.has_lights())
+        return;
+
+    Spectrum extinction_coef = m_volume.extinction_coefficient(
+        m_volume_data, m_volume_ray);
+
+    if (m_distance_sample_count > 0)
+    {
+        const size_t light_count = m_light_sampler.get_non_physical_light_count();
+        
+        // Add contributions from non-physical light sources that don't belong to the lightset.
+        for (size_t light_idx = 0; light_idx < light_count; ++light_idx)
+        {
+            LightSample light_sample;
+            m_light_sampler.sample_non_physical_light(m_time, light_idx, light_sample);
+
+            for (size_t i = 0; i < 2 * m_distance_sample_count; ++i)
+            {
+                add_single_distance_sample_contribution_exponential_only(
+                    &light_sample,
+                    extinction_coef,
+                    sampling_context,
+                    mis_heuristic,
+                    radiance,
+                    false);
+            }
+        }
+    }
+
+    // Add contributions from the light set.
+    if (m_light_sampler.has_lightset())
+    {
+        for (size_t i = 0; i < 2 * m_distance_sample_count; ++i)
+        {
+            for (size_t j = 0; j < m_light_sample_count; ++j)
```

Comment:
So we're taking `2 * m_light_sample_count * m_distance_sample_count` distance samples when using exponential sampling only...